### DISCUSSION TITLE
Simpler backlinks

### DISCUSSION THREAD

**Discussion by alpcantal** [Zettlr/Zettlr#2572] [Ideas]
(NOTE: There are 2 TL;DR versions found at the latter parts)

Workflow wise, when a "filter when following links" is implemented, then three things are supposed to happen:

- filtering when following links helps implement [this feature](https://github.com/Zettlr/Zettlr/issues/2443) without strictly limiting the feature to YAML headers
- following links can now be translated into filtering filenames. Friction when following links can be eliminated by automatically switching the note with closest matching terms in the filename, a la notational velocity or Resophnotes. (I don't know about the code) 
- when clicking on a note ID, the filtered results will already show the backlinks, i.e. the notes that mention the ID. (Automatically generated backlinks lack context, anyway, so **there's NO need to automate putting backlinks into the content area** — it's basically the same as they've showed up in the filtered results). I mentioned this because of [this discussion](https://github.com/Zettlr/Zettlr/discussions/2410). EDIT: [This is already in the roadmap](https://github.com/Zettlr/Zettlr/issues/1522), but I suggest that "filter when following links" is a better alternative than backlinks as a sidebar, for reasons described below.

Another benefit is that many users of Notational Velocity or The Archive cannot find any alternative to those apps in Windows. But this might be a "small probability" benefit.

In my experience I believe Zettlr is pretty much the closest thing to The Archive in Windows, with more features. (I personally love how it integrated pandoc) But instead of searching, filtering must be used to achieve the same (or better) performance.

### Putting it to the test, some problems were revealed

I have tested the "filter" workflow and it [the filter] seems to read plain-text tags very well, but again, not the content area, thus being very good for searching regularly formatted zettel. (more on this in a bit)

So I try to follow links just by using the filter. Here's what I did:

1. Double click on link
2. Copy
3. Ctrl + Shift + T
4. Paste
5. Press "Down" key
6. Press "Enter"

![aFKesZIIdr](https://user-images.githubusercontent.com/64248029/136136085-8a0facf8-eb6e-4446-b05c-fa57f11c0aff.gif)

**One problem that may be encountered by using _just the filter_ is the possibility that other notes are visible**. While quick, there's not enough visual hierarchy for the filter results — hence the need to "follow the link through a filename". But I argue that there is no need for visual hierarchy in filter search results, when there is **automatic switching to note** with closest matching filename. (as mentioned earlier)

Next I try filtering using my own workflow of creating special tags and bookmarks, which I heavily use search for:

![image](https://user-images.githubusercontent.com/64248029/136136354-5e04c75a-f20f-48b9-bc2d-c13580fb9c54.png)

The results showed up instantly! It's excellent when searching for characters in headers. _Very_ excellent.

Next, the bookmarks — I use "qqq" as a bookmark for "where I left off", but then I also use other bookmarks like "✱✱" or "••" and map them in my autohotkey to create a "DIY hotkey-based saved search," but I digress. Here's the screenshots:

![image](https://user-images.githubusercontent.com/64248029/136134392-b4ae3326-33ec-4dbc-a918-6dcb89c454ee.png)

![image](https://user-images.githubusercontent.com/64248029/136134398-d1bbc2b5-bbcf-45ed-a0cb-06deba83e858.png)

**Thus, the _workflow_ caveats**: 

- Plain-text bookmarking (or, "QQ technique") and "searching" in the Zettelkasten Method workflow can be done only through tagging near the note title when using the filter feature. 
- Thus, keyword searching _should_ be done through full-text, which is fine because that's what it's supposed to do.

The problem here, with both of the caveats above, is when you consider the backlinks, but I'll expand more later. 

To expand on the workflow, "filter" function works very well for searching ID's and tags — which makes it a great candidate for a zettelkasten that has notes formatted as follows:

1. Header (ID, Title, Tags)
2. Content (body of note)
3. Footnotes (comments/references, if any)

...which is a well-accepted way of formatting a "Zettel", in my perspective.

## Backlinks posed a dilemma...

The real caveat, however, is that the **backlinks** will NOT show up **unless the filter feature can search for ID's inside the note — which, of course, requires searching the content**. I don't know how code the code works, but this is what it requires at a very high-level. 

Simply put, it's "just" combining full-text search and filtering... 

The dilemma is that filter doesn't search everything, and _that's why_ it's fast. But filter results cannot show backlinks automatically if it cannot search everything!

So, can the filter be "coded" such that it can also search for bracketed terms inside other notes? But that would require searching the entire note, anyway, right? Thus I think the answer to this is to **integrate full-text search into the filter feature, but limit it STRICTLY to the current directory** to retain speed. This way, we have a way of doing incremental full-text search. 

It's what made The Archive (open source), a popular Zettelkasten app for MacOS, frictionless, in my experience. I've used sublimeless_zk, The Archive, Obsidian, ResophNotes, but hey, _Zettlr is open-source_ and available for Windows. (heavy breathing) But I digress.

Now, for those who desperately need backlinks, (if there are indeed people like that...) of course there is a workaround for the current versions.

**Backlinks workaround**: My workaround here is to just do a fulltext search of the ID to show "backlinks". Still, there's also a problem of lack of visual hierarchy (as in bolded/highlighted matching terms), but that's very irrelevant because you're just searching for possibly overlooked connections, not a unique note. In _my_ practice, useful backlinks turn out to be few.

The problem with the current workaround is, of course, the friction. You can embed it into your habits, but it's pretty handy to see backlinks every time you see a current note.

This is what brings me to my summary.

## TL;DR - The dilemma, summary, & practical implication...

**Function**: 
- following links puts the query in the "filter" function, and this automatically switches to the note with the closest matching file name. As a result of this, filter results _automatically show backlinks — which happens even as you switch from note to note_ (See [this video for a sample how it looks like](https://youtu.be/4DIEiIH-N24?t=184))

**Problem**: 
- filter currently lacks the ability to search for ID's/links _inside_ the other notes, and thus cannot show notes that do have backlinks as filter results.
- there is no "filter when following links" function yet — but there is already a "search when following links" function, so I'm guessing this is a possible candidate for a feature

**Current workarounds to implement [Function]**:
- Following a link using the regular Zettlr follow link feature, in case you have a Zettelkasten with a uniform ID scheme 
- Filtering using the ID if you have other existing ID's that don't match the current convention, say you have _both_ YYYYMMDDhhmm and YYYYMMDDhhmmss
- For backlinks, searching for backlinks manually using Ctrl + Shift + F and then pasting the ID (which is slow)

**Therefore, if filtering could at least search for bracketed terms inside other notes, then the [function] stated above would combine the [current methods] into a _single click_ without needing any extraneous features for "automatic backlinks". Thus it's 100% pragmatic for the Zettelkasten Method** so long as you use ID's rather than [ID + filename] as links. (which is the default way in Zettlr)

**Benefits**: 
- As they are simple features, they don't deviate from plain-text approach nor create feature-dependence (i.e. everything mentioned has an _existing_ workaround — the feature set is just a way to eliminate a LOT of friction)
- NO need to automate putting backlinks into the content area — which has the _possibility_ of damaging the contents of the notes
- These simple features would _automatically_ implement [this](https://github.com/Zettlr/Zettlr/issues/1522) part of the roadmap (automatic backlinks) without additional work
- Goes hand-in-hand with [this part of the roadmap](https://github.com/Zettlr/Zettlr/issues/1283), (keyboard switching)

Also, visual hierarchy (e.g. bolded matching search terms) in filter results will NOT be necessary (but still great!) because of automatic switching.

_This is an edited version of my issue posted in https://github.com/Zettlr/Zettlr/issues/2560#issuecomment-935343081_

## Some other thoughts

The problem I see here is my suggested features becomes redundant with the "double bracket" way of linking to other notes. Still, I think it makes following links faster _if note switching_ is used. 

It may also be in conflict in the way Zettlr opens new tabs when opening files. 

I don't have any ideas about these, since I'm not a programmer.

If you want to see a workflow that's similar to what I'm saying, you only have to look at the people who teach _and actually practice_ the Zettelkasten Method: https://www.youtube.com/watch?t=2267

![image](https://user-images.githubusercontent.com/64248029/136027463-c6e8a621-30f5-46fd-9bdc-0bdb26da00a5.png)

Disclaimer: I've been a Zettlr user since early 2020, and I want to contribute potentially helpful ideas to help support the development of Zettlr. I honestly think these ideas are far-fetched, but I've already toned them down into more plausible ones based on my experience using the Zettelkasten Method and being mentored by the guys at zettelkasten.de.

**Comment by alpcantal** [comment]
TL;DR version:

- My idea is to integrate filter (because of the incremental search) and content searching. The difference here is that when using Filter, the results only show _note titles_ that contain the search terms — similar to its original behavior — rather than the exact passages INSIDE filenames, in the case of actual full-text search. So it's more like "Show me the files that contain [query]" rather than "Find every file and line that contains [query] and show me how that passage looks like with [query]" (in the case of the current full-text search).
- I also state the benefits above, especially the fact that simple full-text filtering, if turned into full-text incremental search, _automatically_ shows backlinks (which I speculate is highly requested) without having to implement any weird impractical graphs or backlink sections. When implemented with other features, namely, "_filter when following links_" and "_automatically **switch** to note with _closest_ matching filename_" — it allows a Zettelkasten user to switch from note to note and see backlinks as search results in a single click rather than having them in multiple processes. (high friction! 😫) Basically my ideas were behaviors inspired by nValt and The Archive, applied to Zettlr's Filter function.
- The fact that Zettlr opens files in new tabs may be in conflict with this the features above. But that's just a hypothesis. I don't know how many more tabs it can handle right now that it's in 2.x :)

**Comment by nathanlesage** [comment]
Thanks for opening the discussion! Let me quickly add my two cents from past experience + me as a developer here.

1. First of all, would you mind trying out your ideas with the novel betas? I'm recommending these to everyone basically, because they are stable right now; I'm just waiting for a few other minor things before going stable. This would be important info to know, since some things might've changed in between 1.8.9 and 2.0.0!
2. Then, technically the differences between full-text-search and filtering is as follows: On startup, Zettlr reads in every file and constructs a file cache containing the extracted metadata. Among these are tags, the document's ID (I tend to speak of documents rather than notes or "Zettels" but they're technically the same) and other generic stuff. This is what the filter searches and this is wha makes it faster than full-text search, because the data is already in memory. It's a trade-off: We can't store full files in the cache since that would make the cache unnecessary, but metadata can still be powerful. However, this being said, the file cache is basically also a full-text search which is only executed once instead of every time you hit "search". And as such, we can add whatever additional data we wish to that search. Among that will be in the future links to other documents within each file. Thus, backlink search is effectively easy to implement.
3. Technically, it is a no-brainer to populate the filter field rather than executing a full-text search. It's as simple as adding an option to choose what to do when following a link.
4. With regard to the tab behavior: That's one of the things 2.0.0 does way better.

So much for now, feel free to ask if anything is unclear!

**Reply by alpcantal** [reply]
Thanks for sharing in detail!

>     1. First of all, would you mind trying out your ideas with the novel betas? I'm recommending these to everyone basically, because they are stable right now; I'm just waiting for a few other minor things before going stable. This would be important info to know, since some things might've changed in between 1.8.9 and 2.0.0!

Sorry for that! My screenshots above were taken from 1.8.9, but I've tested them both in 2.0 beta 5. The functions showed the same results:

![image](https://user-images.githubusercontent.com/64248029/136156146-e2416c47-ffcb-49bd-892b-d2964bdb5f05.png)

![image](https://user-images.githubusercontent.com/64248029/136156245-187caebe-0679-4d34-b3a1-bbc37e437ed0.png)

![image](https://user-images.githubusercontent.com/64248029/136156188-8c93869c-b710-497e-9508-94616d8e7f08.png)

Some changes I felt:

The filtering felt blazing fast in the novel versions, I can say! 

The "switching" function was actually extremely fast, too! I'm surprised I didn't notice this at first. Clicking on a note in the directory doesn't open new tabs, so I'd think the feature set I mentioned above just became even more feasible. (EDIT: I noticed there was a new option in the Preferences specifically for this one. Awesome)

>     2. Then, technically the differences between full-text-search and filtering is as follows: On startup, Zettlr reads in every file and constructs a file cache containing the extracted metadata. Among these are tags, the document's ID (I tend to speak of documents rather than notes or "Zettels" but they're technically the same) and other generic stuff. This is what the filter searches and this is wha makes it faster than full-text search, because the data is already in memory. It's a trade-off: We can't store full files in the cache since that would make the cache unnecessary, but metadata can still be powerful. However, this being said, the file cache is basically also a full-text search which is only executed once instead of every time you hit "search". And as such, we can add whatever additional data we wish to that search. Among that will be in the future links to other documents within each file. Thus, backlink search is effectively easy to implement.

That's amazing! Then the tags could be placed anywhere in the documents and still get caught by the filter. :) To clarify, does this mean it's possible to perform a full-text search for document links only, so that we can avoid storing full files (and making the cache unnecessary)?

>     3. Technically, it is a no-brainer to populate the filter field rather than executing a full-text search. It's as simple as adding an option to choose what to do when following a link.

That's great to know. To summarize, based on the current features/capabilities and what you mentioned so far, then the only things needed to implement the idea above are:

- Include caching for document ID's on startup
- Option to filter when following links
- ~~**Switching to a note with the closest matching search term _while filtering_?** → in my perspective, this is _ideal_ for eliminating friction since we're following links often, BUT the next best alternative would be bolding/highlighting of filter results for visual hierarchy.~~ **EDIT**: Explained in the next GitHub comment below

From this idea, the _necessity_ for a "window/pane history" would be eliminated as well because the backlinks already show up in the filter results.

>     4. With regard to the tab behavior: That's one of the things 2.0.0 does way better.

100% agree. I now feel like I'm using a browser when using Zettlr tabs 😄 Zettlr has come a long way!! Kudos to you, man.

**Reply by alpcantal** [reply]
Upon reflection, my thinking here was actually extraneous — the third feature I proposed was unnecessary because I (stupidly) realized we can already follow links in Zettlr. 

**The most minimal implementation would be ONLY these two**:

- Include caching for document ID's on startup
- Option to filter when following links

Should I turn these into their own GitHub issues?? :)

**Reply by nathanlesage** [reply]
Feel free to! Won't come directly in 2.0 but this way it's on the radar! But what you want to do is instead of wording it "Include caching for document IDs" please word it like so: "Include file references (forward-links) in the FSAL metadata". Then we have it there, and I just have to add it to the fields the filter will check.

**Reply by alpcantal** [reply]
Done! Good to know it'll be on the radar. Not a programmer, but since the features are simple and useful enough I would be hopeful for it to come quickly after you finish some bugs :D Thanks for this discussion, I'm glad to have contributed my ideas coming from my experience in other similar apps.

**Reply by alpcantal** [reply]
Hey man @nathanlesage, is this still on your radar? Thanks
